### ISSUE TITLE
[Suggestion] Support to color

### ISSUE THREAD

**Issue by cualquiercosa327** [skyfloogle/red-viper#44]
Hello, first I want to give thanks for your awesome work : it is a awesome way to could enjoy this underrated machine.

I want to suggest the option of making it could use different colors as it does some gb emulators.

[colorize your games](http://www.mobile-dev.ch/old.php?page=pcsoft_heigboy)

It would awesome user could do "patches" and maybe add this feature. 
Maybe it could be implemented on a way than homebrew developers could use it when compilate their stuff allowing keep the colors on a external patch and their binary go on full 100% virtual boy compatibility. (I suppose it would be the part where the sprites are converted to vb format when the extra data would keep on the ips file).
Thanks

**Comment by DuplexEagle** [comment]
This would especially be awesome since the Virtual Boy was originally envisioned by Nintendo to have a full colour display.

**Comment by cualquiercosa327** [comment]
> This would especially be awesome since the Virtual Boy was originally envisioned by Nintendo to have a full colour display.

Oh, I didn't know that.
Thanks for sharing.

**Comment by eadmaster** [comment]
related implementations in other emus: https://github.com/SupernaviX/vvb/issues/79

**Comment by cualquiercosa327** [comment]
Hi. I had no seen that.
It looks nice but it seems hard to build the color files.
Also it would be great if more than 4 color could be used (as we
are talking of an emulator, it should no be harder.
As example (Although it was more complex, it was done by Sega with their "Gigadrive" for 3ds

**Comment by Cyber-Crud** [comment]
I completely agree.  The VB basically has a 4-color palette of red, similar to how the GameBoy worked. Thus the 4-color palette options from the GBC or the Super GameBoy on the SNES.  

There's no reason that couldn't be implemented here.  The default palette for the Super GameBoy was strong.  I've looked at the emulator source code and it does appear that the hardware has the 4 color hex values for the emulated red display.   This could easily either be hard-coded with new palette values or made into a variable to make palette swaps modular on the system.  

The roms don't contain the color values because they're basically B&W on a red LED.  So the emulation is just using a 4 color red palette -- basically a solid red, light red, magenta, and black.  There's no reason this couldn't be modified.  I'd hate to have to fork this just for a palette change when it can just be easily added to this project. 

I think 4-color palettes can be done right now and give us Super GameBoy level colonization.  Beyond that, anything else would require some significant work.  And the 3DS doesn't have glsl shader support like RetroArch does on Windows.  So we're basically stuck with what we can do.

**Comment by Cyber-Crud** [comment]
The Super GameBoy took what was essentially a monochromatic GameBoy game and added 2-bit color to it. And just that little change made a world of difference.  Not only does it help eye fatigue, but your brain will actually seemingly fill in even more colors than the 4 that are actually showing.  

If you consider the GameBoy having just a green LED the same way that the VirtualBoy has a red one, then you can see how going from this:

<img width="1080" height="764" alt="Image" src="https://github.com/user-attachments/assets/e2d57368-ce7d-4041-b5a3-8c01d0d15c59" />

...to this:

<img width="823" height="720" alt="Image" src="https://github.com/user-attachments/assets/34b3e108-37de-4edc-9ca7-dc3d1f2a44e3" />

...would do absolute wonders for VB roms.  We're basically recycling late 90s tech with late 90s tech. It's literally the same idea.

**Comment by skyfloogle** [comment]
> I've looked at the emulator source code and it does appear that the hardware has the 4 color hex values for the emulated red display.
When I load Virtual Boy tiles into the 3DS's GPU, I convert the three possible colour values to red, green, and blue (the fourth is transparent). When I draw tiles, I multiply each component by a new colour depending on the palette (VB internal palettes, so another black/red/green/blue) and add them together. During post-processing, I multiply each component by a new brightness value (the red values are configurable - this is used during fading effects) and add them together and tint with the configured tint colour. This system is entirely reliant on the final image being monochrome.

Those hex values you found are probably transparent (or black), red, green, and blue. Changing them would break the whole system.

In any case, I am currently working on retooling the graphics system to allow for 4 separate hues. I can't simply do the same thing during post processing that I do for tiles, because the brightness values can actually change across the screen (see the intro cinematic in Wario Land). I'm investigating a solution where the colour information is provided by three different sources (I don't have room in the pipeline for three more textures).

**Comment by Cyber-Crud** [comment]
Awesome, thanks for the explanation.  I'm glad you're trying to figure it out.  The fact that you have a color wheel that allows the base color to change from red to basically any other hue, is already a huge improvement on eye fatigue.  

I assumed of the VB utilizes a similar 4-color palette, even if it's 4 magnitudes of brightness, then that could be converted to 4 separate colors rather than 1 with different hues.  Might be doable.  

Thanks for replying.

**Comment by skyfloogle** [comment]
Implemented as of 4a88fe0. Will be included in a release soon.

**Comment by skyfloogle** [comment]
When 3D mode is already active, I can't say having different colours per eye sounds particularly appealing to look at.
However, there is a customizable anaglyph mode already present, if you're interested in that sort of thing.

**Comment by Cyber-Crud** [comment]
Haha yeah I hear ya.  But future rom hacks and homebrews could utilize that to make some crazy effects.  At the very least, the possibility is already there.  

Very excited for this update.  It's going to make everyone flip out!

**Comment by Cyber-Crud** [comment]
![Image](https://github.com/user-attachments/assets/8194afbf-f16d-4199-837d-f664ae75a97c)

This is the biggest day in gaming history!  Legit 4-color 3D VirtualBoy!!!!
